pyfr/plugins/turbulence.py

```diff
@@ -0,0 +1,272 @@
+            ostatet *= mcgmultiplier
+            ostatet ^= ostatet >> n22
+            random = (ostatet >> 8) * pow(2, -24)
+            yield (ostate, random)
+
+
+class TurbulencePlugin(BaseSolverPlugin):
+    name = 'turbulence'
+    systems = ['ac-navier-stokes', 'navier-stokes']
+    formulations = ['dual', 'std']
+    dimensions = [3]
+
+    def __init__(self, intg, cfgsect):
+        super().__init__(intg, cfgsect)
+
+        self.fdptype = fdptype = intg.backend.fpdtype
+
+        ac = intg.system.name.startswith('ac')
+
+        self.nmvxinit = 6
+        self.nvmxtol = 0.1
+
+        self.tstart = intg.tstart
+        self.tbegin = intg.tcurr
+        self.tnext = intg.tcurr
+        self.tend = intg.tend
+
+        self.seed = self.cfg.getint(cfgsect, 'seed')
+
+        self.eventdtype = [('tinit', fdptype), ('state', np.uint32),
+                           ('ts', fdptype), ('te', fdptype)]
+
+        ti = self.cfg.getfloat(cfgsect, 'turbulence-intensity')
+        sigma = self.cfg.getfloat(cfgsect, 'sigma')
+        self.avgu = avgu = self.cfg.getfloat(cfgsect, 'avg-u')
+        self.ls = ls = self.cfg.getfloat(cfgsect, 'turbulence-length-scale')
+
+        gc = (2*sigma/(math.erf(1/sigma)*math.pi**0.5))**0.5
+        beta1 = -0.5/(sigma*ls)**2
+        beta3 = 0.01*ti*avgu*(gc/sigma)**3
+
+        if not ac:
+            gamma = self.cfg.getfloat('constants', 'gamma')
+            avgrho = self.cfg.getfloat(cfgsect, 'avg-rho')
+            avgmach = self.cfg.getfloat(cfgsect, 'avg-mach')
+            beta2 = avgrho*(gamma - 1)*avgmach**2
+        else:
+            beta2 = 0
+
+        self.ydim = ydim = self.cfg.getfloat(cfgsect, 'y-dim')
+        self.zdim = zdim = self.cfg.getfloat(cfgsect, 'z-dim')
+
+        self.ymin = ymin = -ydim/2
+        self.zmin = zmin = -zdim/2
+
+        self.nvorts = int(ydim*zdim/(4*ls**2))
+        self.bbox = BoxRegion([-2*ls, ymin-ls, zmin-ls],
+                              [2*ls, ymin+ydim+ls, zmin+zdim+ls])
+
+        centre = self.cfg.getliteral(cfgsect, 'centre')
+        rax = np.array(self.cfg.getliteral(cfgsect, 'rot-axis'))
+        ran = np.radians(self.cfg.getfloat(cfgsect, 'rot-angle'))
+
+        self.shift = shift = np.array(centre)
+        self.rot = rot = (np.cos(ran)*np.eye(3) +
+                          np.sin(ran)*(np.cross(rax, np.eye(3))) +
+                          (1 - np.cos(ran))*rax @ rax.T)
+
+        self.macro_params = {'ls': ls, 'avgu': avgu, 'yzmin': [ymin, zmin],
+                             'yzdim': [ydim, ydim], 'beta1': beta1,
+                             'beta2': beta2, 'beta3': beta3, 'rot': rot,
+                             'shift': shift, 'ac': ac}
+
+        self.trcl = defaultdict()
+        self.vortbuf = self.getvortbuf()
+        self.vortstructs = self.getvortstructs(intg)
+
+        if not bool(self.vortstructs):
+            self.tnext = float('inf')
+
+    def __call__(self, intg):
+        if intg.tcurr + intg._dt < self.tnext:
```

Comment:
We typically add a tolerance when calling plugins based on time. Do you think it's worth adding?

Reply:
I don't think its needed here. I think that's more relevant when we need the plugin to do something at a particular time. Here we just need to make sure that the buffer of info on the device side never runs dry, and the if statement that is there should do that.